text/2019-02-26-tikv-client-command-line.md

```diff
@@ -0,0 +1,208 @@
+be very easy to issue one-off commands to TiKV.
+
+Further, we mentioned in the Rust Client RFC
+(https://github.com/tikv/rfcs/pull/7) we mentioned we wanted to accomplish this.
+
+## Detailed design
+
+The client will be created using the `tikv-client` crate and its already existing
+functionality. It will utilize `clap` to provide a simple wrapper around the
+crate API.
+
+### Install
+
+We will build this tool in the `tikv-cli` repository. It will be published.
+Users will be able to install it with `cargo install tikv-cli`. (See
+*Alternatives* for another option.)
+
+### Configuration
+
+The tool will retrieve its configuration from one of the following places, in
+order of precedence (greatest to least):
+
+* The command line options, such as `--pd 127.0.0.1:2379`.
+* A local configuration file `./.tikv-cli.toml`
+* The global configuration file `${XDG_CONFIG_HOME}/tikv-cli.toml`.
+* The default options (connect to localhost, no TLS).
+
+The options are:
+
+```toml
+[connection]
+# A list of `pd` endpoints.
+pd_endpoints = [
+    "127.0.0.1:2379",
+    "127.0.0.2:2379",
+]
+
+[tls]
+# The path of the CA file for TLS.
+ca_path = "./rsa.ca"
+# The path of the certificate file for TLS.
+cert_path = "./rsa.cert"
+# The path of the key file for TLS.
+key_path = "./rsa.key"
+
+[output]
+# If the ouput JSON should be minified.
+minify = false
+# Level of logging to use. Available: error. warn, info, debug, trace
+logging = "info"
+```
+
+### Input/Output
+
+By default input is by positional arguments in `bash`/`sh` string format.
+
+By default the tool will output in prettyprinted JSON. For single entry results,
+it will return a string (including `"`), for key-value pairs it will use a JSON
+object `{ key: "", value: "" }`, for multiple results it will return a JSON
+array. Errors will use an `{ error: "" }` object. Opting out of prettyprint will
+be via the `--minify` flag.
+
+JSON was chosen because it is an efficient, pervasive, human-readable, modern
+standard that has many tools (eg `jq`) to use for post processing if needed.
+
+All standard (non-logging) output will be on `stdout`.
+
+It will feature configurable logging levels with `slog` which will output to `stderr`.
+
+The tool will exit zero on successful requests, and non-zero on unsuccessful
+requests. In the case of REPL mode, it shall always exit zero.
+
+### Interacting with it
+
+Users will interact with the tool through the command line. It will support both
+a one-off mode or a REPL (Read-eval-print-loop) style mode (See alternatives).
+
+Example of the one-off raw mode:
+
+```bash
+$ tikv-cli raw set "cncf" "linux foundation"
+Finished in 0.001s.
+$ tikv-cli raw set "tikv" "rust"
+Finished in 0.001s.
+$ tikv-cli raw get "cncf"
+"linux foundation"
+Finished in 0.002s.
+$ tikv-cli raw scan "b"..
+[
+    { key: "cncf", value: "linux foundation" },
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s.
+```

```

Comment:
These "Finished in 0.003s.", will they be on stderr or stdout? Are they configurable?

Reply:
I think these should be stderr.

Reply:
Added a switch for these